File: myfactory/commands/orchestrator.py
"""`myfactory orchestrator prompt` - print the Hermes controller prompt.

Prefers the project's own copy under .ApplicationFactory/orchestrator/;
falls back to the core template so the command works before init.
"""

from __future__ import annotations

import argparse

from myfactory.core import paths

PROMPT_FILENAME = "HERMES-CONTROLLER-PROMPT.md"


def register(subparsers: argparse._SubParsersAction) -> None:
    p = subparsers.add_parser(
        "orchestrator",
        help="Hermes orchestrator/controller helpers.",
        description="Hermes controls sprint execution: it delegates to agents and enforces QA gates.",
    )
    sub = p.add_subparsers(dest="orchestrator_command", metavar="SUBCOMMAND")

    prompt = sub.add_parser(
        "prompt",
        help="Print the Hermes controller prompt.",
        description=(
            "Prints the prompt to give the Hermes controller agent. Uses the "
            "project's .ApplicationFactory/orchestrator/ copy when present, "
            "otherwise the core template."
        ),
    )
    prompt.add_argument("--target", default=None, help="Target directory (default: current directory).")
    prompt.set_defaults(func=run_prompt)

    p.set_defaults(func=lambda args: (p.print_help(), 0)[1])


def run_prompt(args: argparse.Namespace) -> int:
    target = paths.resolve_target(args.target)

    project_prompt = paths.orchestrator_dir(target) / PROMPT_FILENAME
    template_prompt = (
        paths.product_template_dir() / ".ApplicationFactory" / "orchestrator" / PROMPT_FILENAME
    )

    if project_prompt.is_file():
        source = project_prompt
        origin = "project"
    elif template_prompt.is_file():
        source = template_prompt
        origin = "core template (project not initialized - run `myfactory init`)"
    else:
        print("ERROR: Hermes controller prompt not found in project or core templates.")
        return 1

    print(f"# Source: {origin}: {source}")
    print()
    print(source.read_text(encoding="utf-8"))
    return 0

File: internal/commands/orchestrator.go
// Command orchestrator prompt: print the Hermes controller prompt.
// Prefers the project's own copy under .ApplicationFactory/orchestrator/;
// falls back to the embedded template so the command works before init.
package commands

import (
	"flag"
	"fmt"
	"io"
	"path/filepath"

	"github.com/AlharbiKhalid/myfactory-core/internal/assets"
	"github.com/AlharbiKhalid/myfactory-core/internal/fsops"
	"github.com/AlharbiKhalid/myfactory-core/internal/project"
)

const promptFilename = "HERMES-CONTROLLER-PROMPT.md"

// Orchestrator implements `myfactory orchestrator`.
func Orchestrator(args []string, stdout, stderr io.Writer) int {
	if len(args) == 0 || args[0] == "-h" || args[0] == "--help" {
		fmt.Fprintln(stdout, "usage: myfactory orchestrator prompt [flags]")
		fmt.Fprintln(stdout)
		fmt.Fprintln(stdout, "Hermes controls sprint execution: it delegates to agents and enforces QA gates.")
		fmt.Fprintln(stdout)
		fmt.Fprintln(stdout, "subcommands:")
		fmt.Fprintln(stdout, "  prompt    Print the Hermes controller prompt.")
		return 0
	}
	if args[0] != "prompt" {
		fmt.Fprintf(stderr, "ERROR: unknown orchestrator subcommand %q (expected: prompt)\n", args[0])
		return 2
	}
	return orchestratorPrompt(args[1:], stdout, stderr)
}

func orchestratorPrompt(args []string, stdout, stderr io.Writer) int {
	fl := flag.NewFlagSet("orchestrator prompt", flag.ContinueOnError)
	fl.SetOutput(stderr)
	target := fl.String("target", "", "Target directory (default: current directory).")
	if err := fl.Parse(args); err != nil {
		return 2
	}
	targetDir, err := project.ResolveTarget(*target)
	if err != nil {
		fmt.Fprintf(stderr, "ERROR: %v\n", err)
		return 1
	}
	// The embedded fallback exists so the command works before init, but an
	// explicitly requested target that does not exist is a user error and
	// must not silently fall back.
	if *target != "" {
		if err := project.EnsureDir(targetDir); err != nil {
			fmt.Fprintf(stderr, "ERROR: %v\n", err)
			return 1
		}
	}

	projectPrompt := filepath.Join(project.OrchestratorDir(targetDir), promptFilename)
	if fsops.FileExists(projectPrompt) {
		content, err := fsops.ReadFileString(projectPrompt)
		if err != nil {
			fmt.Fprintf(stderr, "ERROR: %v\n", err)
			return 1
		}
		fmt.Fprintf(stdout, "# Source: project: %s\n\n", projectPrompt)
		fmt.Fprintln(stdout, content)
		return 0
	}

	content, err := assets.ReadFile("product-repo/.ApplicationFactory/orchestrator/" + promptFilename)
	if err != nil {
		fmt.Fprintln(stdout, "ERROR: Hermes controller prompt not found in project or embedded templates.")
		return 1
	}
	fmt.Fprintf(stdout, "# Source: embedded template (project not initialized - run `myfactory init`)\n\n")
	fmt.Fprintln(stdout, string(content))
	return 0
}
